dbtmetabase/models/interface.py

```diff
@@ -0,0 +1,127 @@
+
+from .config import MetabaseConfig, DbtConfig
+from .metabase import MetabaseModel
+from .exceptions import (
+    NoDbtPathSupplied,
+    NoDbtSchemaSupplied,
+    MetabaseClientNotInstantiated,
+    DbtParserNotInstantiated,
+    MetabaseUnableToSync,
+)
+from ..parsers.dbt_folder import DbtFolderReader
+from ..parsers.dbt_manifest import DbtManifestReader
+from ..metabase import MetabaseClient
+
+
+class MetabaseInterface(MetabaseConfig):
+    """Interface for interacting with instantiating a Metabase Config and preparing a client object"""
+
+    _client: Optional[MetabaseClient] = None
+
+    @property
+    def client(self) -> MetabaseClient:
+        if self._client is None:
+            raise MetabaseClientNotInstantiated(
+                "Metabase client is not yet instantiated. Call `prepare_metabase_client` method first"
+            )
+        return self._client
+
+    def prepare_metabase_client(self, dbt_models: Optional[List[MetabaseModel]] = None):
+        """Prepares the metabase client which can then after be accessed via the `client` property
+
+        Args:
+            dbt_models (Optional[List[MetabaseModel]]): Used if sync is enabled to verify all dbt models passed exist in Metabase
+
+        Raises:
+            MetabaseUnableToSync: This error is raised if sync is enabled and a timeout is explicitly set in the `Metabase` object config
+        """
+        if dbt_models is None:
+            dbt_models = []
+
+        self._client = MetabaseClient(
+            host=self.host,
+            user=self.user,
+            password=self.password,
+            use_http=self.use_http,
+            verify=self.verify,
+        )
+
+        # Sync and attempt schema alignment prior to execution; if timeout is not explicitly set, proceed regardless of success
+        if self.sync:
+            if self.sync_timeout is not None and not self._client.sync_and_wait(
+                self.database,
+                dbt_models,
+                self.sync_timeout,
+            ):
+                logging.critical("Sync timeout reached, models still not compatible")
+                raise MetabaseUnableToSync(
+                    "Unable to align models between dbt target models and Metabase"
+                )
+
+
+class DbtInterface(DbtConfig):
+    """Interface for interacting with instantiating a Dbt Config and preparing a validated parser object"""
+
+    _parser: Optional[Union[DbtManifestReader, DbtFolderReader]] = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.validate_config()
+        self.prepare_dbt_parser()
+
+    @property
+    def parser(self) -> Union[DbtManifestReader, DbtFolderReader]:
+        if self._parser is None:
+            raise DbtParserNotInstantiated(
+                "dbt reader is not yet instantiated. Call `prepare_dbt_parser` method first"
+            )
+        return self._parser
+
+    def get_config(self) -> DbtConfig:
```

Comment:
If `DbtInterface` class extends `DbtConfig`, you can just use the former as the latter. Or am I missing something?

Reply:
Yes but I though instantiating an interface to expose a parser and then passing the object (self) reference to that parser seemed a little circular. If I were going to do that, I'd just make a method that internally uses the parser and passes the self ref. 

This way is more verbose, but makes it clearer that the config (a DbtConfig) is all that is required to use the function --not the DbtInterface...
Does that make sense, what do you think?

Reply:
I'd do the internal passing thing. Otherwise, that single use of it looks confusing - you're basically passing a representation of something to a function of a member of that something. That member (parser) should just be set up with a reference back to its owner (interface). 

For lifecycle purposes, interface still owns parser, but parser is just made aware of what owns it. Similar to how a child node in a graph implementation might have a reference back to its parent, in case it needs to propagate back up for any reason.

Reply:
Ok I added a method to the interface which passes its self ref to the function and users can use that instead of accessing the raw uninformed parser.
Definitely looks cleaner and makes more sense. Thank you